File: Python/面试经典150题/回溯/39_组合总和.py
from typing import List
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        l = len(candidates)
        def dfs(arr, now, pos):
            if now == target:
                ans.append(arr[:])
                return
            if now > target:
                return
            for i in range(pos,l):
                arr.append(candidates[i])
                now += candidates[i]
                dfs(arr, now, i)
                arr.pop()
                now -= candidates[i]
        dfs([],0,0)
        return ans

print(Solution().combinationSum([2,3,6,7],7))

File: JS/回溯算法/组合/39_组合总和.js
var combinationSum = function(candidates, target) {
    let result = [];
    const find = (arr, sum, now) => {
        if(sum === target)    {
            result.push(arr)
            return
        }
        if(sum > target)    {
            return
        }
        for(let i=now; i<candidates.length; i++){
            let newArr = [...arr]
            newArr.push(candidates[i])
            find(newArr, sum+candidates[i], i)
        }
    }
    find([], 0, 0)
    return result
};
console.log(combinationSum([2,3,6,7], 7));